runtime/strings/strings.cpp

```diff
@@ -0,0 +1,291 @@
+#include<stdlib.h>
+#include<string.h>
+#include<algorithm>
+#include<cstdint>
+#include<iomanip>
+#include<string>
+#include<sstream>
+
+extern "C" {
+
+#define KCHAR char
+
+struct blockheader {
+	int64_t len;
+};
+
+struct string {
+	blockheader b;
+	KCHAR data[0];
+};
+
+bool hook_STRING_gt(const string * a, const string * b) {
+	auto res = memcmp(a->data, b->data, std::min(a->b.len, b->b.len));
+	return res > 0 || (res == 0 && a->b.len > b->b.len);
+}
+
+bool hook_STRING_gte(const string * a, const string * b) {
+	auto res = memcmp(a->data, b->data, std::min(a->b.len, b->b.len));
+	return (res > 0 || (res == 0 && a->b.len >= b->b.len));
+}
+
+bool hook_STRING_lt(const string * a, const string * b) {
+	auto res = memcmp(a->data, b->data, std::min(a->b.len, b->b.len));
+	return res < 0 || (res == 0 && a->b.len < b->b.len);
+}
+
+bool hook_STRING_lte(const string * a, const string * b) {
+	auto res = memcmp(a->data, b->data, std::min(a->b.len, b->b.len));
+	return (res < 0 || (res == 0 && a->b.len <= b->b.len));
+}
+
+bool hook_STRING_eq(const string * a, const string * b) {
+	auto res = memcmp(a->data, b->data, std::min(a->b.len, b->b.len));
+	return (res == 0 && a->b.len == b->b.len);
+}
+
+bool hook_STRING_ne(const string * a, const string * b) {
+	auto res = memcmp(a->data, b->data, std::min(a->b.len, b->b.len));
+	return (res != 0 || a->b.len != b->b.len);
+}
+
+string * hook_STRING_concat(const string * a, const string * b) {
+	auto len_a = a->b.len;
+	auto len_b = b->b.len;
+	auto newlen = len_a  + len_b;
+	auto ret = static_cast<string *>(malloc(sizeof(string) + newlen));
+	ret->b.len = newlen;
+	memcpy(&(ret->data), &(a->data), a->b.len * sizeof(KCHAR));
+	memcpy(&(ret->data[a->b.len]), &(b->data), b->b.len * sizeof(KCHAR));
+	return ret;
+}
+
+int64_t hook_STRING_length(const string * a) {
+	return a->b.len;
+}
+
+string * hook_STRING_chr(const int64_t ord) {
+	auto ret = static_cast<string *>(malloc(sizeof(string) + sizeof(KCHAR)));
+	ret->b.len = 1;
+	ret->data[0] = static_cast<KCHAR>(ord);
+	return ret;
+}
+
+int64_t hook_STRING_ord(const string * input) {
```

Comment:
similarly, this returns an mpz_ptr which you should malloc and then call mpz_init_set_ui. Note that you should cast to an unsigned type rather than a signed type, otherwise if you're on a system where `char` is signed, you might return a negative number from this function.